### ISSUE TITLE
LSP goto definition not working on macOS

### ISSUE THREAD

**Issue by kevinsjoberg** [vamolessa/pepper#27]
I'm not getting goto definition to work. See the video below in which I'm trying to use `gd` to go to definition for the word under the cursor in the pepper codebase.

[![asciicast](https://asciinema.org/a/Ue6pk7rQJU2QD1jfOCLNOten3.svg)](https://asciinema.org/a/Ue6pk7rQJU2QD1jfOCLNOten3)

I've confirmed that LSP is working since `K` (for hover) does work and do give me information about the word under the cursor.

**Environment**

OS: macOS 12.0.1
Terminal: Kitty 0.24.2
pepper: 0.22.0

**Comment by kevinsjoberg** [comment]
Here's the LSP log for reference:

<details>
<summary>LSP log</summary>

```console
send request
method: 'initialize'
params:
{"processId":45908,"clientInfo":{"name":"pepper-plugin-lsp","version":"0.10.0"},"rootUri":"file:////Users/kevinsjoberg/code/pepper","capabilities":{"workspace":{"applyEdit":true,"workspaceFolders":false,"configuration":false,"didChangeWatchedFiles":{},"executeCommand":{},"workspaceEdit":{"documentChanges":true,"failureHandling":"undo","resourceOperations":["create","rename","delete"],"changeAnnotationSupport":null},"symbol":{"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}}},"textDocument":{"synchronization":{"willSave":false,"willSaveWaitUntil":false,"didSave":true},"completion":{"contextSupport":false,"completionItem":{"snippetSupport":false,"commitCharactersSupport":false,"documentationFormat":["plaintext"],"deprecatedSupport":false,"preselectSupport":false,"tagSupport":{"valueSet":[1]}},"completionItemKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]}},"hover":{"contentFormat":["plaintext"]},"signatureHelp":{"contextSupport":false,"signatureInformation":{"documentationFormat":["plaintext"],"parameterInformation":{"labelOffsetSupport":false}}},"declaration":{"linkSupport":false},"definition":{"linkSupport":false},"typeDefinition":{"linkSupport":false},"implementation":{"linkSupport":false},"references":{},"documentSymbol":{"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]},"hierarchicalDocumentSymbolSupport":true,"labelSupport":false},"codeAction":{"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}}},"documentLink":{"tooltipSupport":false},"formatting":{},"rangeFormatting":{},"rename":{"prepareSupport":true,"prepareSupportDefaultBehavior":1,"honorsChangeAnnotations":false},"publishDiagnostics":{"relatedInformation":true,"tagSupport":{"valueSet":[1]},"versionSupport":false},"selectionRange":{}},"window":{"workDoneProgress":false,"showMessage":{},"showDocument":{"support":true}},"offsetEncoding":["utf-8"]}}

receive response
id: 1
method: 'initialize'
result:
{"capabilities":{"textDocumentSync":{"openClose":true,"change":2,"save":{}},"selectionRangeProvider":true,"hoverProvider":true,"completionProvider":{"triggerCharacters":[":",".","'"]},"signatureHelpProvider":{"triggerCharacters":["(",","]},"definitionProvider":true,"typeDefinitionProvider":true,"implementationProvider":true,"referencesProvider":true,"documentHighlightProvider":true,"documentSymbolProvider":true,"workspaceSymbolProvider":true,"codeActionProvider":{"codeActionKinds":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite"],"resolveProvider":true},"codeLensProvider":{"resolveProvider":true},"documentFormattingProvider":true,"documentRangeFormattingProvider":false,"documentOnTypeFormattingProvider":{"firstTriggerCharacter":"=","moreTriggerCharacter":[".",">","{"]},"renameProvider":{"prepareProvider":true},"foldingRangeProvider":true,"declarationProvider":true,"workspace":{"fileOperations":{"willRename":{"filters":[{"scheme":"file","pattern":{"glob":"**/*.rs","matches":"file"}},{"scheme":"file","pattern":{"glob":"**","matches":"folder"}}]}}},"callHierarchyProvider":true,"semanticTokensProvider":{"legend":{"tokenTypes":["comment","keyword","string","number","regexp","operator","namespace","type","struct","class","interface","enum","enumMember","typeParameter","function","method","property","macro","variable","parameter","angle","arithmetic","attribute","attributeBracket","bitwise","boolean","brace","bracket","builtinAttribute","builtinType","character","colon","comma","comparison","constParameter","derive","dot","escapeSequence","formatSpecifier","generic","label","lifetime","logical","macroBang","operator","parenthesis","punctuation","selfKeyword","semicolon","typeAlias","toolModule","union","unresolvedReference"],"tokenModifiers":["documentation","declaration","definition","static","abstract","deprecated","readonly","defaultLibrary","async","attribute","callable","constant","consuming","controlFlow","crateRoot","injected","intraDocLink","library","mutable","public","reference","trait","unsafe"]},"range":true,"full":{"delta":true}},"experimental":{"joinLines":true,"openCargoToml":true,"ssr":true,"onEnter":true,"parentModule":true,"hoverRange":true,"runnables":{"kinds":["cargo"]},"workspaceSymbolScopeKindFiltering":true}},"serverInfo":{"name":"rust-analyzer","version":"8e9ccbf97 2022-01-03 stable"},"offsetEncoding":"utf-8"}

send notification
method: 'initialized'
params:
{}

send notification
method: 'textDocument/didOpen'
params:
{"textDocument":{"uri":"file:////Users/kevinsjoberg/code/pepper/./pepper/src/command/builtin.rs","languageId":"rust","version":1,"text":"use std::path::Path;\n\nuse crate::{\n    buffer::{parse_path_and_position, BufferProperties},\n    buffer_position::BufferPosition,\n    client::ViewAnchor,\n    command::{CommandError, CommandIO, CommandManager, CompletionSource},\n    config::{ParseConfigError, CONFIG_NAMES},\n    cursor::Cursor,\n    editor::{EditorContext, EditorFlow},\n    editor_utils::MessageKind,\n    help,\n    mode::{picker, read_line, ModeKind},\n    syntax::TokenKind,\n    theme::{Color, THEME_COLOR_NAMES},\n};\n\npub fn register_commands(commands: &mut CommandManager) {\n    let mut r = |name, completions, command_fn| {\n        commands.register(None, name, completions, command_fn);\n    };\n\n    static HELP_COMPLETIONS: &[CompletionSource] = &[CompletionSource::Commands];\n    r(\"help\", HELP_COMPLETIONS, |ctx, io| {\n        let keyword = io.args.try_next();\n        io.args.assert_empty()?;\n\n        let client_handle = io.client_handle()?;\n        let (path, position) = match keyword.and_then(help::search) {\n            Some((path, position)) => (path, position),\n            None => (help::main_help_name(), BufferPosition::zero()),\n        };\n\n        let mut buffer_path = ctx.editor.string_pool.acquire();\n        buffer_path.push_str(help::HELP_PREFIX);\n        buffer_path.push_str(path);\n        match ctx.editor.buffer_view_handle_from_path(\n            client_handle,\n            Path::new(&buffer_path),\n            BufferProperties::scratch(),\n            true,\n        ) {\n            Ok(handle) => {\n                let client = ctx.clients.get_mut(client_handle);\n                client.set_buffer_view_handle(Some(handle), &ctx.editor.buffer_views);\n                client.set_view_anchor(&ctx.editor, ViewAnchor::Center);\n\n                let mut cursors = ctx.editor.buffer_views.get_mut(handle).cursors.mut_guard();\n                cursors.clear();\n                cursors.add(Cursor {\n                    anchor: position,\n                    position,\n                });\n            }\n            Err(error) => ctx\n                .editor\n                .status_bar\n                .write(MessageKind::Error)\n                .fmt(format_args!(\"{}\", error)),\n        }\n        ctx.editor.string_pool.release(buffer_path);\n\n        Ok(())\n    });\n\n    r(\"print\", &[], |ctx, io| {\n        let mut write = ctx.editor.status_bar.write(MessageKind::Info);\n        if let Some(arg) = io.args.try_next() {\n            write.str(arg);\n        }\n        while let Some(arg) = io.args.try_next() {\n            write.str(\"\\n\");\n            write.str(arg);\n        }\n        Ok(())\n    });\n\n    r(\"quit\", &[], |ctx, io| {\n        io.args.assert_empty()?;\n        if ctx.clients.iter().count() == 1 {\n            io.assert_can_discard_all_buffers(ctx)?;\n        }\n        io.flow = EditorFlow::Quit;\n        Ok(())\n    });\n\n    r(\"quit-all\", &[], |ctx, io| {\n        io.args.assert_empty()?;\n        io.assert_can_discard_all_buffers(ctx)?;\n        io.flow = EditorFlow::QuitAll;\n        Ok(())\n    });\n\n    r(\"open\", &[CompletionSource::Files], |ctx, io| {\n        let path = io.args.next()?;\n\n        let mut properties = BufferProperties::text();\n        while let Some(property) = io.args.try_next() {\n            match property {\n                \"text\" => properties = BufferProperties::text(),\n                \"scratch\" => properties = BufferProperties::scratch(),\n                \"history-enabled\" => properties.history_enabled = true,\n                \"history-disabled\" => properties.history_enabled = false,\n                \"saving-enabled\" => properties.saving_enabled = true,\n                \"saving-disabled\" => properties.saving_enabled = false,\n                \"word-database-enabled\" => properties.word_database_enabled = true,\n                \"word-database-disabled\" => properties.word_database_enabled = false,\n                _ => return Err(CommandError::NoSuchBufferProperty),\n            }\n        }\n\n        let client_handle = io.client_handle()?;\n        let (path, position) = parse_path_and_position(path);\n\n        let path = Path::new(&path);\n        match ctx.editor.buffer_view_handle_from_path(\n            client_handle,\n            Path::new(path),\n            properties,\n            true,\n        ) {\n            Ok(handle) => {\n                let client = ctx.clients.get_mut(client_handle);\n                client.set_buffer_view_handle(Some(handle), &ctx.editor.buffer_views);\n\n                if let Some(position) = position {\n                    let mut cursors = ctx.editor.buffer_views.get_mut(handle).cursors.mut_guard();\n                    cursors.clear();\n                    cursors.add(Cursor {\n                        anchor: position,\n                        position,\n                    });\n                }\n            }\n            Err(error) => ctx\n                .editor\n                .status_bar\n                .write(MessageKind::Error)\n                .fmt(format_args!(\"{}\", error)),\n        }\n\n        Ok(())\n    });\n\n    r(\"save\", &[CompletionSource::Files], |ctx, io| {\n        let path = io.args.try_next().map(|p| Path::new(p));\n        io.args.assert_empty()?;\n\n        let buffer_handle = io.current_buffer_handle(ctx)?;\n        let buffer = ctx.editor.buffers.get_mut(buffer_handle);\n\n        buffer\n            .write_to_file(path, &mut ctx.editor.events)\n            .map_err(CommandError::BufferWriteError)?;\n\n        ctx.editor\n            .status_bar\n            .write(MessageKind::Info)\n            .fmt(format_args!(\"buffer saved to {:?}\", &buffer.path));\n        Ok(())\n    });\n\n    r(\"save-all\", &[], |ctx, io| {\n        io.args.assert_empty()?;\n\n        let mut count = 0;\n        for buffer in ctx.editor.buffers.iter_mut() {\n            if buffer.properties.saving_enabled {\n                buffer\n                    .write_to_file(None, &mut ctx.editor.events)\n                    .map_err(CommandError::BufferWriteError)?;\n                count += 1;\n            }\n        }\n\n        ctx.editor\n            .status_bar\n            .write(MessageKind::Info)\n            .fmt(format_args!(\"{} buffers saved\", count));\n        Ok(())\n    });\n\n    r(\"reopen\", &[], |ctx, io| {\n        io.args.assert_empty()?;\n\n        let buffer_handle = io.current_buffer_handle(ctx)?;\n        io.assert_can_discard_buffer(ctx, buffer_handle)?;\n        let buffer = ctx.editor.buffers.get_mut(buffer_handle);\n\n        buffer\n            .read_from_file(&mut ctx.editor.word_database, &mut ctx.editor.events)\n            .map_err(CommandError::BufferReadError)?;\n\n        ctx.editor\n            .status_bar\n            .write(MessageKind::Info)\n            .str(\"buffer reopened\");\n        Ok(())\n    });\n\n    r(\"reopen-all\", &[], |ctx, io| {\n        io.args.assert_empty()?;\n\n        io.assert_can_discard_all_buffers(ctx)?;\n        let mut count = 0;\n        for buffer in ctx.editor.buffers.iter_mut() {\n            buffer\n                .read_from_file(&mut ctx.editor.word_database, &mut ctx.editor.events)\n                .map_err(CommandError::BufferReadError)?;\n            count += 1;\n        }\n\n        ctx.editor\n            .status_bar\n            .write(MessageKind::Info)\n            .fmt(format_args!(\"{} buffers reopened\", count));\n        Ok(())\n    });\n\n    r(\"close\", &[], |ctx, io| {\n        io.args.assert_empty()?;\n\n        let buffer_handle = io.current_buffer_handle(ctx)?;\n        io.assert_can_discard_buffer(ctx, buffer_handle)?;\n        ctx.editor\n            .buffers\n            .defer_remove(buffer_handle, &mut ctx.editor.events);\n\n        ctx.editor\n            .status_bar\n            .write(MessageKind::Info)\n            .str(\"buffer closed\");\n\n        Ok(())\n    });\n\n    r(\"close-all\", &[], |ctx, io| {\n        io.args.assert_empty()?;\n\n        io.assert_can_discard_all_buffers(ctx)?;\n        let mut count = 0;\n        for buffer in ctx.editor.buffers.iter() {\n            ctx.editor\n                .buffers\n                .defer_remove(buffer.handle(), &mut ctx.editor.events);\n            count += 1;\n        }\n\n        ctx.editor\n            .status_bar\n            .write(MessageKind::Info)\n            .fmt(format_args!(\"{} buffers closed\", count));\n        Ok(())\n    });\n\n    static CONFIG_COMPLETIONS: &[CompletionSource] = &[CompletionSource::Custom(CONFIG_NAMES)];\n    r(\"config\", CONFIG_COMPLETIONS, |ctx, io| {\n        let key = io.args.next()?;\n        let value = io.args.try_next();\n        io.args.assert_empty()?;\n\n        match value {\n            Some(value) => match ctx.editor.config.parse_config(key, value) {\n                Ok(()) => Ok(()),\n                Err(error) => Err(CommandError::ConfigError(error)),\n            },\n            None => match ctx.editor.config.display_config(key) {\n                Some(display) => {\n                    ctx.editor\n                        .status_bar\n                        .write(MessageKind::Info)\n                        .fmt(format_args!(\"{}\", display));\n                    Ok(())\n                }\n                None => Err(CommandError::ConfigError(ParseConfigError::NoSuchConfig)),\n            },\n        }\n    });\n\n    static COLOR_COMPLETIONS: &[CompletionSource] = &[CompletionSource::Custom(THEME_COLOR_NAMES)];\n    r(\"color\", COLOR_COMPLETIONS, |ctx, io| {\n        let key = io.args.next()?;\n        let value = io.args.try_next();\n        io.args.assert_empty()?;\n\n        let color = ctx\n            .editor\n            .theme\n            .color_from_name(key)\n            .ok_or(CommandError::NoSuchColor)?;\n\n        match value {\n            Some(value) => {\n                let encoded =\n                    u32::from_str_radix(value, 16).map_err(|_| CommandError::NoSuchColor)?;\n                *color = Color::from_u32(encoded);\n            }\n            None => ctx\n                .editor\n                .status_bar\n                .write(MessageKind::Info)\n                .fmt(format_args!(\"0x{:0<6x}\", color.into_u32())),\n        }\n\n        Ok(())\n    });\n\n    r(\"map-normal\", &[], |ctx, io| map(ctx, io, ModeKind::Normal));\n    r(\"map-insert\", &[], |ctx, io| map(ctx, io, ModeKind::Insert));\n    r(\"map-command\", &[], |ctx, io| {\n        map(ctx, io, ModeKind::Command)\n    });\n    r(\"map-readline\", &[], |ctx, io| {\n        map(ctx, io, ModeKind::ReadLine)\n    });\n    r(\"map-picker\", &[], |ctx, io| map(ctx, io, ModeKind::Picker));\n\n    static ALIAS_COMPLETIONS: &[CompletionSource] =\n        &[CompletionSource::Custom(&[]), CompletionSource::Commands];\n    r(\"alias\", ALIAS_COMPLETIONS, |ctx, io| {\n        let from = io.args.next()?;\n        let to = io.args.next()?;\n        io.args.assert_empty()?;\n        ctx.editor.commands.aliases.add(from, to);\n        Ok(())\n    });\n\n    r(\"syntax\", &[], |ctx, io| {\n        let glob = io.args.next()?;\n        io.args.assert_empty()?;\n        match ctx.editor.syntaxes.set_current_from_glob(glob) {\n            Ok(()) => Ok(()),\n            Err(error) => Err(CommandError::InvalidGlob(error)),\n        }\n    });\n\n    r(\"syntax-keywords\", &[], |ctx, io| {\n        syntax_pattern(ctx, io, TokenKind::Keyword)\n    });\n    r(\"syntax-types\", &[], |ctx, io| {\n        syntax_pattern(ctx, io, TokenKind::Type)\n    });\n    r(\"syntax-symbols\", &[], |ctx, io| {\n        syntax_pattern(ctx, io, TokenKind::Symbol)\n    });\n    r(\"syntax-literals\", &[], |ctx, io| {\n        syntax_pattern(ctx, io, TokenKind::Literal)\n    });\n    r(\"syntax-strings\", &[], |ctx, io| {\n        syntax_pattern(ctx, io, TokenKind::String)\n    });\n    r(\"syntax-comments\", &[], |ctx, io| {\n        syntax_pattern(ctx, io, TokenKind::Comment)\n    });\n    r(\"syntax-texts\", &[], |ctx, io| {\n        syntax_pattern(ctx, io, TokenKind::Text)\n    });\n\n    r(\"copy-command\", &[], |ctx, io| {\n        let command = io.args.next()?;\n        io.args.assert_empty()?;\n        ctx.platform.copy_command.clear();\n        ctx.platform.copy_command.push_str(command);\n        Ok(())\n    });\n\n    r(\"paste-command\", &[], |ctx, io| {\n        let command = io.args.next()?;\n        io.args.assert_empty()?;\n        ctx.platform.paste_command.clear();\n        ctx.platform.paste_command.push_str(command);\n        Ok(())\n    });\n\n    r(\"enqueue-keys\", &[], |ctx, io| {\n        let keys = io.args.next()?;\n        io.args.assert_empty()?;\n\n        ctx.editor\n            .buffered_keys\n            .parse(keys)\n            .map_err(|e| CommandError::KeyParseError(e.error))?;\n        Ok(())\n    });\n\n    r(\"find-file\", &[], |ctx, io| {\n        let command = io.args.next()?;\n        let prompt = io.args.try_next().unwrap_or(\"open:\");\n        io.args.assert_empty()?;\n        picker::find_file::enter_mode(ctx, command, prompt);\n        Ok(())\n    });\n\n    r(\"find-pattern\", &[], |ctx, io| {\n        let command = io.args.next()?;\n        let prompt = io.args.try_next().unwrap_or(\"find:\");\n        io.args.assert_empty()?;\n        read_line::find_pattern::enter_mode(ctx, command, prompt);\n        Ok(())\n    });\n\n    r(\"pid\", &[], |ctx, io| {\n        io.args.assert_empty()?;\n        ctx.editor\n            .status_bar\n            .write(MessageKind::Info)\n            .fmt(format_args!(\"{}\", std::process::id()));\n        Ok(())\n    });\n}\n\nfn map(ctx: &mut EditorContext, io: &mut CommandIO, mode: ModeKind) -> Result<(), CommandError> {\n    let from = io.args.next()?;\n    let to = io.args.next()?;\n    io.args.assert_empty()?;\n\n    match ctx.editor.keymaps.parse_and_map(mode, from, to) {\n        Ok(()) => Ok(()),\n        Err(error) => Err(CommandError::KeyMapError(error)),\n    }\n}\n\nfn syntax_pattern(\n    ctx: &mut EditorContext,\n    io: &mut CommandIO,\n    token_kind: TokenKind,\n) -> Result<(), CommandError> {\n    let pattern = io.args.next()?;\n    io.args.assert_empty()?;\n    match ctx\n        .editor\n        .syntaxes\n        .get_current()\n        .set_rule(token_kind, pattern)\n    {\n        Ok(()) => Ok(()),\n        Err(error) => Err(CommandError::PatternError(error)),\n    }\n}"}}

receive notification
method: 'textDocument/publishDiagnostics'
params:
{"uri":"file:///Users/kevinsjoberg/code/pepper/pepper/src/platforms/unix_utils.rs","diagnostics":[{"range":{"start":{"line":75,"character":35},"end":{"line":75,"character":41}},"severity":2,"code":"deprecated","source":"rustc","message":"use of deprecated function `libc::daemon`: Deprecated in MacOSX 10.5\n`#[warn(deprecated)]` on by default","tags":[2]}]}

receive notification
method: 'textDocument/publishDiagnostics'
params:
{"uri":"file:///Users/kevinsjoberg/code/pepper/pepper/src/platforms/unix_utils.rs","diagnostics":[]}

receive notification
method: 'textDocument/publishDiagnostics'
params:
{"uri":"file:///Users/kevinsjoberg/code/pepper/pepper/src/platforms/unix_utils.rs","diagnostics":[{"range":{"start":{"line":75,"character":35},"end":{"line":75,"character":41}},"severity":2,"code":"deprecated","source":"rustc","message":"use of deprecated function `libc::daemon`: Deprecated in MacOSX 10.5\n`#[warn(deprecated)]` on by default","tags":[2]}]}

receive notification
method: 'textDocument/publishDiagnostics'
params:
{"uri":"file:///Users/kevinsjoberg/code/pepper/pepper/src/command/builtin.rs","diagnostics":[],"version":1}

send request
method: 'textDocument/definition'
params:
{"textDocument":{"uri":"file:////Users/kevinsjoberg/code/pepper/./pepper/src/command/builtin.rs"},"position":{"line":56,"character":17}}

receive response
id: 2
method: 'textDocument/definition'
result:
[{"uri":"file:///Users/kevinsjoberg/code/pepper/pepper/src/editor.rs","range":{"start":{"line":278,"character":8},"end":{"line":278,"character":18}}}]

send notification
method: 'textDocument/didOpen'
params:
{"textDocument":{"uri":"file:////Users/kevinsjoberg/code/pepper/Users/kevinsjoberg/code/pepper/pepper/src/editor.rs","languageId":"rust","version":1,"text":""}}

send notification
method: 'textDocument/didClose'
params:
{"textDocument":{"uri":"file:////Users/kevinsjoberg/code/pepper/Users/kevinsjoberg/code/pepper/pepper/src/editor.rs"}}
````
</details>

I think there's something funky going on with the URI. Look at the URI for the last `textDocument/didOpen` call.

**Comment by vamolessa** [comment]
I've just noticed I'm getting the same error on linux as well. I really does seem like something wrong is happening when dealing with paths from lsp.

I suspect it's the part I try to remove the base path from the returned uri in order to keep a clean filepath.
Gonna investigate.

**Comment by vamolessa** [comment]
I think this should be fixed on 47c89502
Please reopen if there's still something wrong happening
